### PR TITLE
Rustdoc : stop cloning html render context

### LINKED ISSUE #82381: rustdoc: Get rid of shared mutable state in `Context`
`Context` currently has a [`.shared`] field, which is shared mutable state.
It would be nice if we could redesign the code such that we don't have shared
mutable state.

One idea I have for accomplishing this is passing `&mut SharedContext` in
wherever it's needed, instead of getting it out of `Context`.

[`.shared`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/html/render/struct.Context.html#structfield.shared

### PR DESCRIPTION
**Pull request by Wardenfar**
The issue was about cleaning the shared context in html rendering.
The problem was that the Context was clone for each item rendered.

I change the rendering process to not clone the context so the shared context do not require anymore to be in a std::rc::Rc

Feel free to give me feedback
This PR may impact performance: should we run perf-test ? 

Resolves #82381

### CODE REVIEW

File: src/librustdoc/formats/renderer.rs [hunk 1]
```diff
@@ -65,33 +72,64 @@ pub(crate) fn run_format<'tcx, T: FormatRenderer<'tcx>>(
...
+                } else {
+                    // the last item of the module has been rendered
+                    // -> exit the module
+                    format_renderer.mod_item_out()?;
+                }
             }
-
-            cx.mod_item_out()?;
-        // FIXME: checking `item.name.is_some()` is very implicit and leads to lots of special
-        // cases. Use an explicit match instead.
```
Comment (jyn514) [inline src/librustdoc/formats/renderer.rs:89]: You removed this comment. Please add it back.

**Comment by Wardenfar** [comment]
Do I have to do something ?

**Comment by camelid** [comment]
Thanks for doing this! I will do my best to look at soon. :)

**Comment by camelid** [comment]
I was able to look at this today, and overall it looks very good! I will give it a full review within a few days.

**Review by camelid — CHANGES_REQUESTED** [review]
Sorry for taking so long to review this!

My one concern is that you added several `RefCell`s and `Cell`s to the new `ContextItemFrame` struct and changed `&mut Context` to `&Context` everywhere. I don't see why this is necessary, and we've been trying really hard to move away from interior mutability. Is it possible to directly mutate `ContextItemFrame` through `&mut Context` instead?

If that's not possible, please let me know and we'll figure out how to go forward.

**Comment by Wardenfar** [comment]
@camelid 
In a lot of places, when removing `RefCell<Shared>` the code no longer compiles with the borrow checker.
In the rendering/printing process, only the `ContextItemFrame` is mutated. To remove completely `RefCell` / `Cell`, you need to remove the `ContextItemFrame` from the `Context` struct (to allow `&Context` and `&mut ContextItemFrame` at the same time)

I tried today but `Context` is everywhere. This is why i choose to keep `RefCell` (but in a smaller part of the struct)

The other way is to rewrite places when the borrow checker fail, but this is a lot of time.

**Comment by camelid** [comment]
Thanks for your reply!

Why do we need `&Context` and `&mut ContextItemFrame` simultaneously? Can't we have `&mut Context` and use that to get the item frames? I haven't looked at this code in a while, so apologies if I'm missing something.

**Comment by JohnCSimon** [comment]
@Wardenfar 
Ping from triage: 
Can  you address the comment from the reviewer and fix the merge conflicts?

FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by Wardenfar** [comment]
@JohnCSimon 
Sorry I don't have time to work on it
Should I close the PR ?

**Comment by JohnCSimon** [comment]
> @JohnCSimon Sorry I don't have time to work on it Should I close the PR ?

I'll close it for you. 
Please reopen when you are ready to continue with this. 
Note: if you do please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**